crates/ruff_python_formatter/src/expression/string/docstring.rs

```diff
@@ -702,19 +784,336 @@ impl<'src> CodeExampleDoctest<'src> {
+        // > (together called the "directive marker"). Directive types
+        // > are case-insensitive single words (alphanumerics plus
+        // > isolated internal hyphens, underscores, plus signs, colons,
+        // > and periods; no whitespace).
+        //
+        // The language names matched here (e.g., `python` or `py`) are taken
+        // from the [Pygments lexer names], which is referenced from the docs
+        // for the [code-block] directive.
+        //
+        // [directives]: https://docutils.sourceforge.io/docs/ref/rst/restructuredtext.html#directives
+        // [Pygments lexer names]: https://pygments.org/docs/lexers/
+        // [code-block]: https://www.sphinx-doc.org/en/master/usage/restructuredtext/directives.html#directive-code-block
+        static DIRECTIVE_START: Lazy<Regex> = Lazy::new(|| {
+            Regex::new(
+                r"(?m)^\s*\.\. \s*(?i:code-block|sourcecode)::\s*(?i:python|py|python3|py3)$",
+            )
+            .unwrap()
+        });
+        if !DIRECTIVE_START.is_match(original.line) {
+            return None;
+        }
+        Some(CodeExampleRst {
+            lines: vec![],
+            opening_indent: indentation_length(original.line),
+            min_indent: None,
+            is_directive: true,
+        })
+    }
+
+    /// Returns the code collected in this example as a sequence of lines.
+    ///
+    /// The lines returned have the minimum indentation stripped from their
+    /// prefix. Based on the definition of minimum indentation, this implies
+    /// there is at least one line in the slice returned with no whitespace
+    /// prefix.
+    fn code(&mut self) -> &[CodeExampleLine<'src>] {
+        let Some(min_indent) = self.min_indent else {
+            return &[];
+        };
+        for line in &mut self.lines {
+            line.code = if line.original.line.trim().is_empty() {
+                ""
+            } else {
+                indentation_trim(min_indent, line.original.line)
+            };
+        }
+        &self.lines
+    }
+
+    /// Attempts to add the given line from a docstring to the reStructuredText
+    /// code snippet being collected.
+    ///
+    /// This takes ownership of `self`, and if ownership is returned to the
+    /// caller, that means the caller should continue trying to add lines to
+    /// this code snippet. Otherwise, if ownership is not returned, then this
+    /// implies at least one action was added to the give queue to either reset
+    /// the code block or format. That is, the code snippet was either found to
+    /// be invalid or it was completed and should be reformatted.
+    ///
+    /// Note that actions may be added even if ownership is returned. For
+    /// example, empty lines immediately preceding the actual code snippet will
+    /// be returned back as an action to print them verbatim, but the caller
+    /// should still continue to try to add lines to this code snippet.
+    fn add_code_line(
```

Comment:
What happens with empty lines at the end of a docstring (with a unclosed example above). Does it get trimmed by the formatter?

Reply:
Nice test case. I added one:

```python
# If a literal block is never properly ended (via a non-empty unindented line),
# then the end of the block should be the last non-empty line. And subsequent
# empty lines should be preserved as-is.
def rst_literal_extra_blanks_at_end():
    """
    Do cool stuff::


        cool_stuff( 1 )



    """
    pass
```

In this case, just the code snippet should get reformatted to `cool_stuff(1)`, but the empty lines are preserved as they are not part of the code snippet. The reST renderer I'm using seems to agree.

Ah, but I figured you were noticing something awry, so I revisited the logic with this prompt in mind and indeed, a bug has been found! Namely, in this code:

```
# A literal block can contain many empty lines and it should not end the block
# if it continues.
def rst_literal_extra_blanks_in_snippet():
    """
    Do cool stuff::

        cool_stuff( 1 )


        cool_stuff( 2 )

    Done.
    """
    pass
```

`cool_stuff( 1 )` will get formatted, but `cool_stuff( 2 )` will not because the block is prematurely closed.

OK, fixed that while not regressing the first test.